pkg/ingester/ingester.go

```diff
@@ -470,6 +502,16 @@ func (i *Ingester) starting(ctx context.Context) (err error) {
 			_ = services.StopAndAwaitTerminated(context.Background(), i.lifecycler)
 		}
 	}()
+
+	// First of all we have to check if the shutdown marker is set. This needs to be done
+	// as first thing because, if found, it may change the behaviour of the ingester startup.
+	if exists, err := shutdownmarker.Exists(shutdownmarker.GetPath(i.cfg.BlocksStorageConfig.TSDB.Dir)); err != nil {
+		return errors.Wrap(err, "failed to check ingester shutdown marker")
+	} else if exists {
+		level.Info(i.logger).Log("msg", "detected existing shutdown marker, setting unregister and flush on shutdown", "path", shutdownmarker.GetPath(i.cfg.BlocksStorageConfig.TSDB.Dir))
+		i.setPrepareShutdown()
+	}
```

Comment:
Related to comment on dskit about `SetCreatePartitionOnStartup`, we could check for shutdown marker in `New` function, and create correct config for partition lifecycler there.

Reply:
Briefly discussed in dskit PR. I would keep it as is for now, but I agree with you that could be improved.